pkg/store/bucket.go

```diff
@@ -1544,7 +1564,20 @@ func (r *bucketIndexReader) fetchPostings(groups []*postingGroup) error {
 				// Return postings and fill LRU cache.
 				// Truncate first 4 bytes which are length of posting.
 				groups[p.groupID].Fill(p.keyID, newBigEndianPostings(pBytes[4:]))
-				r.block.indexCache.StorePostings(r.ctx, r.block.meta.ULID, groups[p.groupID].keys[p.keyID], pBytes)
+
+				storeData := pBytes
+
+				if r.block.enablePostingsCompression {
+					// Reencode postings before storing to cache. If that fails, we store original bytes.
+					// This can only fail, if postings data was somehow corrupted,
+					// and there is nothing we can do about it. It's not worth reporting here.
+					data, err := diffVarintSnappyEncode(newBigEndianPostings(pBytes[4:]))
+					if err == nil {
+						storeData = data
+					}
```

Comment:
What if cache is malformed and will give error?

Reply:
Here we are encoding postings data fetched from remote storage. If that data is somehow corrupted (`newBigEndianPostings` return error, or postings entries are not in increasing order), we will store original data (as fetched from storage) to the cache. This is what code did already before this PR.

When same corrupted data is later used, queries will return with errors.

Reply:
Fair :+1: